### PR TITLE
[single] Added API and skeleton for flexible input dimension

### PR DESCRIPTION
**Pull request by kparichay**
Adding support for flexible input dimension
Added API and skeleton for setting input dimension via single API

Check #1651 for more information

**Self evaluation:**
1. Build test: [ ]Passed [ ]Failed [*]Skipped
2. Run test: [ ]Passed [ ]Failed [* ]Skipped

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: api/capi/include/nnstreamer-single.h [hunk 1]
```diff
@@ -138,6 +138,18 @@ int ml_single_get_input_info (ml_single_h single, ml_tensors_info_h *info);
...
 int ml_single_get_output_info (ml_single_h single, ml_tensors_info_h *info);
 
+/**
+ * @brief Sets the information (tensor dimension, type, name and so on) of required input data for the given model.
+ * @details Note that a model may not support setting such information.
+ * @param[in] single The model handle.
+ * @param[in] info The handle of input tensors information.
+ * @return @c 0 on success. Otherwise a negative error value.
+ * @retval #ML_ERROR_NONE Successful
+ * @retval #ML_ERROR_NOT_SUPPORTED Not supported.
```
Comment (myungjoo) [inline api/capi/include/nnstreamer-single.h]: Let's add more description on this error, ```ML_ERROR_NOT_SUPPORTED```, for application developers:
```
Not supported. This implies that the given model or framework does not support dynamic dimensions. Use get input/output ..... APIs instead for this model and framework.
```

**Comment by myungjoo** [comment]
In later commits, let's think about macros / inline-or-simple functions that provides syntactic sugar for making it easier to use dynamic input dimensions. For example:
```
ml_single_invoke_dynamic ( handle[in], input[in], input_dim[in], output[out], output_dim[out]);
```
, which can be something :
```
#define ml_single_invoke_dynamic( ... )  ml-api; ml-api2; ...
```
or 
a static inline function.

File: api/capi/src/nnstreamer-capi-single.c [hunk 2]
```diff
@@ -615,7 +615,7 @@ ml_single_get_tensors_info (ml_single_h single, gboolean is_input,
 }
 
 /**
- * @brief Gets the type (tensor dimension, type, name and so on) of required input data for the given handle.
+ * @brief Gets the information (tensor dimension, type, name and so on) of required input data for the given handle.
```
Comment (dongju-chae) [inline api/capi/src/nnstreamer-capi-single.c]: Q. the given handle == the given model?

Comment (kparichay) [reply api/capi/src/nnstreamer-capi-single.c]: Yes, as this is single API, one handle corresponds to one (single) model.

**Comment by kparichay** [comment]
> In later commits, let's think about macros / inline-or-simple functions that provides syntactic sugar for making it easier to use dynamic input dimensions. For example:

Yes, will add these. Added as TODO in #1651.

File: api/capi/include/nnstreamer-single.h (hunk 1 above)
Comment (kparichay) [reply api/capi/include/nnstreamer-single.h]: Updated to :
```This implies that the given **~model or~** framework does not support dynamic dimensions. Use get input/output (( fill these in ..... )) APIs instead for **~this model and~** framework.```

There is no way to distinguish if the given input information set by the user is invalid or the model does allow setting dynamic dimensions. So, I think both these cases should be ML_ERROR_INVALID_PARAMETER. However, for framework, we can return ML_ERROR_NOT_SUPPORTED.

**Comment by kparichay** [comment]
V2:
Added more description for retval of ML_ERROR_NOT_SUPPORTED in documentation

**Pull request merged**